python/sglang/backend/openai.py

```diff
@@ -152,6 +154,10 @@ def generate(
                             "Example of adding api speculative execution: @function(num_api_spec_tokens=128)."
                         )
                     prompt = s.messages_
+                    # Open AI model requires function call information to be sent to the model
+                    # along with the prompt.
+                    for function_call in s.function_calls:
+                        prompt.append(function_call)
                 else:
```

Comment:
Restructured the code a little bit based on your suggestions (with some minor tweaks but I can update if you think it's still better to move the function call generate call outside of generate (we will just have a simpler generate call):

Within openai.py
- `build_function_call_messages()`: a new function which builds function call messages. Given function signature is specific to open ai models, keeping the logic to parse inputs and produce function call messages within the backend code.
- `generate()`: Given prompt is local to the `generate()` call, I directly added `function_call_messages` to it so that we can call with function call messages during the current completion call's prompt. The main intuition is to try resuing the generate call logic and it also only appends function call response (comp) without intermediate messages into the final text/messages. 

Within interpreter.py
- Updated `_execute_gen()` logic to include building function call messages if tools are provided, and handle both parallel function calling and non-parallel function calling by either calling `backend.generate` one time for parallel function call supported models, or multiple times if parallel call is not supported.